sentinel/src/schedule/outbound.js

```diff
@@ -0,0 +1,259 @@
+  };
+
+  return auth().then(() => request(sendOptions));
+};
+
+// Collects push configs to attempt out of the task's queue, given a global config
+const findConfigurationsToPush = (config, taskDoc) => {
+  return taskDoc.queue.map(pushName => {
+    return config.find(conf => conf.key === pushName);
+  });
+};
+
+const singlePush = (taskDoc, medicDoc, config) => {
+  const infodoc = configService.getTransitionsLib().infodoc;
+
+  const payload = mapDocumentToOutbound(medicDoc, config);
+  return send(payload, config)
+    .then(() => {
+      // Worked, remove entry from queue and add link to info doc
+      logger.info(`Pushed ${medicDoc._id} to ${config.key}`);
+
+      taskDoc.queue.splice(config.key, 1);
+      if (taskDoc.queue.length === 0) {
+        // Done with this queue completely
+        taskDoc._deleted = true;
+      }
+
+      return db.sentinel.put(taskDoc)
+        .then(({rev}) => {
+          // This works because we are running a single push one at a time and we're passing a
+          // shared reference around (so the next singlePush to run against the same task will have
+          // a reference to this same doc with the updated rev). If we refactor this code to be more
+          // efficient we need to be careful to not cause conflicts against the task document.
+          taskDoc._rev = rev;
+        })
+        .then(() => infodoc.get({id: medicDoc._id, doc: medicDoc}))
+        .then(info => {
+          info.completed_tasks = info.completed_tasks || [];
+          info.completed_tasks.push({
+            type: 'outbound',
+            name: config.key,
+            timestamp: Date.now()
+          });
+          return db.sentinel.put(info);
+        });
+    })
+    .catch(err => {
+      // Failed
+      logger.error(`Failed to push ${medicDoc._id} to ${config.key}: ${err.message}`);
+      logger.error(err);
+
+      // Don't remove the entry from the task's queue so it will be tried again next time
+    });
+};
+
+// Coordinates the attempted pushing of documents that need it
+const execute = () => {
+  const pushConfig = arrayinate(configService.get(CONFIGURED_PUSHES) || {});
+  if (!pushConfig.length) {
+    return Promise.resolve();
+  }
+
+  return queuedTasks()
+  .then(tasks => {
+    const pushes = tasks.reduce((acc, {taskDoc, medicDoc}) => {
+      const pushesForDoc =
+        findConfigurationsToPush(pushConfig, taskDoc)
+          .map(config => ({taskDoc, medicDoc, config}));
+
+      return acc.concat(pushesForDoc);
+    }, []);
+
+    // Attempts each push one by one. Written to be simple not efficient.
+    // There are lots of things we could do to make this faster / less fragile,
+    // such as scoping pushes by domain, as well as writing out successes before
+    // all pushes are complete
+    // For now we presume we aren't going to get much traffic against this and
+    // will probably only be doing one push per schedule call
+    return pushes.reduce(
```

Comment:
What about returning an `async.series` result instead of this slightly hard to read `reduce`? (Consistent with `schedule/index.js`)

Reply:
`async` is a library to manipulate functions that use the older callback pattern, so to use it I would have to rewrite the promises code as callbacks.

We do use reduce pretty frequently to manage promises in this way throughout the codebase. How would you achieve this normally?

Reply:
```for (let push of pushes) {
  await singlePush(push);
}```